File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        result = []
        row, column = len(matrix), len(matrix[0])
        left, right = 0, column-1
        up, down = 0, row-1

        while len(result) < row * column:
            for c in range(left, right+1):
                result.append(matrix[up][c])
            for r in range(up+1, down+1):
                result.append(matrix[r][right])
            
            if up != down:
                for c in range(right-1, left-1, -1):
                    result.append(matrix[down][c])
            if left != right:
                for r in range(down-1, up, -1):
                    result.append(matrix[r][left])
            
            up += 1
            down -= 1
            left += 1
            right -= 1
        
        return result
        
        

File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> v;
        int row = matrix.size();
        int column = matrix[0].size();
        int left = 0, right = column-1;
        int up = 0, down = row-1;

        while(v.size() < row*column){
            // left to right
            for(int c = left; c <= right; c++) v.push_back(matrix[up][c]);
            // up to down
            for(int r = up+1; r <= down; r++) v.push_back(matrix[r][right]);
            // right to left
            if(up != down){
                for(int c = right-1; c >= left; c--) v.push_back(matrix[down][c]);
            }
            // down to up
            if(left != right){
                for(int r = down-1; r > up; r--) v.push_back(matrix[r][left]);
            }
            up++; down--;
            left++; right--;
        }
        return v;
    }
};